Switch to static OpenSSL on Windows

dennisameling opened the pull request:
As discussed in https://github.com/journeyapps/node-sqlcipher/issues/73#issuecomment-783393874. Similar to what the Signal team did in their fork: https://github.com/EvanHahn-Signal/node-sqlcipher/commit/16916949f0c010f6e6d3d5869b10a0ab813eae75

Also includes Windows ARM64 libs and updates @mapbox/node-pre-gyp (they just released v1.0.0) 🚀 

Output of `npm run test` on Windows x64:

```
176 passing (11s)
  1 pending
```

Also closes https://github.com/journeyapps/node-sqlcipher/issues/73

rkistner commented on deps/sqlite3.gyp:
```diff
@@ -112,8 +118,9 @@
               "files": [
-                '<(SHARED_INTERMEDIATE_DIR)/sqlcipher-amalgamation-<@(sqlite_version)/>(openssl_root)/libeay32.dll',
-                '<(SHARED_INTERMEDIATE_DIR)/sqlcipher-amalgamation-<@(sqlite_version)/>(openssl_root)/msvcr120.dll'
+                '<(SHARED_INTERMEDIATE_DIR)/sqlcipher-amalgamation-<@(sqlite_version)/>(openssl_root)/libssl.lib',
+                '<(SHARED_INTERMEDIATE_DIR)/sqlcipher-amalgamation-<@(sqlite_version)/>(openssl_root)/libcrypto.lib',
+                '<(SHARED_INTERMEDIATE_DIR)/sqlcipher-amalgamation-<@(sqlite_version)/>(openssl_root)/ossl_static.pdb'
```
Is this needed, or can we remove this entire action?

AFAIK only the dll's needed to be copied, and the .lib files should now be statically linked as part of the .node file.

dennisameling replied:
Can be removed indeed, done!

rkistner commented on deps/sqlite3.gyp:
```diff
@@ -51,8 +55,10 @@
-            '-lssleay32.lib',
+            '-llibcrypto.lib',
+            '-llibssl.lib',
+            '-lws2_32.lib',
+            '-lcrypt32.lib'
```
Are these two required? Would be good to have a comment on what it's used for / why it's needed.

dennisameling replied:
Not required indeed. Came across them in https://github.com/EvanHahn-Signal/node-sqlcipher/commit/16916949f0c010f6e6d3d5869b10a0ab813eae75 so also added them here, but just tested without and it build/tests without problems 😊

dennisameling replied:
Actually, they _are_ needed for the Electron build to succeed. Otherwise it fails with the error below. Just added a note about that in https://github.com/journeyapps/node-sqlcipher/pull/75/commits/35ac248ccc9608132a89f30761efe8b073fe1a31

```
Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
  nothing.c
  win_delay_load_hook.cc
  nothing.vcxproj -> F:\repos\node-sqlcipher\build\Release\\nothing.lib
  unpack_sqlite_dep
  Copying F:\repos\node-sqlcipher\build\Release\obj/global_intermediate/sqlcipher-amalgamation-3033000/OpenSSL-Win64/libcrypto.lib to F:\repos\node-sqlcipher\build\Release\\libcrypto.lib
          1 file(s) copied.
  Copying F:\repos\node-sqlcipher\build\Release\obj/global_intermediate/sqlcipher-amalgamation-3033000/OpenSSL-Win64/libssl.lib to F:\repos\node-sqlcipher\build\Release\\libssl.lib
          1 file(s) copied.
  Copying F:\repos\node-sqlcipher\build\Release\obj/global_intermediate/sqlcipher-amalgamation-3033000/OpenSSL-Win64/ossl_static.pdb to F:\repos\node-sqlcipher\build\Release\\ossl_static.pdb
          1 file(s) copied.
  sqlite3.c
  win_delay_load_hook.cc
  sqlite3.vcxproj -> F:\repos\node-sqlcipher\build\Release\\sqlite3.lib
  backup.cc
  database.cc
  node_sqlite3.cc
  statement.cc
  win_delay_load_hook.cc
     Creating library F:\repos\node-sqlcipher\build\Release\node_sqlite3.lib and object F:\repos\node-sqlcipher\build\Release\node_sqlite3.exp
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_ioctlsocket [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_getsockname [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_getsockopt [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_getsockopt [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_ntohs [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_addr.obj) : error LNK2001: unresolved external symbol __imp_ntohs [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_gethostbyname [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_WSAStartup [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_WSACleanup [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_WSAGetLastError [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(bss_sock.obj) : error LNK2001: unresolved external symbol __imp_WSAGetLastError [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_WSAGetLastError [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_addr.obj) : error LNK2001: unresolved external symbol __imp_getaddrinfo [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_addr.obj) : error LNK2001: unresolved external symbol __imp_freeaddrinfo [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_addr.obj) : error LNK2001: unresolved external symbol __imp_getnameinfo [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertOpenStore [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertCloseStore [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertEnumCertificatesInStore [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertFindCertificateInStore [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertDuplicateCertificateContext [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertFreeCertificateContext [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(e_capi.obj) : error LNK2001: unresolved external symbol __imp_CertGetCertificateContextProperty [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(bss_sock.obj) : error LNK2001: unresolved external symbol __imp_recv [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(bss_sock.obj) : error LNK2001: unresolved external symbol __imp_send [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(bss_sock.obj) : error LNK2001: unresolved external symbol __imp_WSASetLastError [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_accept [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_bind [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_closesocket [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_connect [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_listen [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_setsockopt [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
libcrypto.lib(b_sock2.obj) : error LNK2001: unresolved external symbol __imp_socket [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
F:\repos\node-sqlcipher\build\Release\node_sqlite3.node : fatal error LNK1120: 28 unresolved externals [F:\repos\node-sqlcipher\build\node_sqlite3.vcxproj]
```

rkistner reviewed (commented):
Thanks for the contribution! I added some minor comments - I just want to make sure we only include what we need.

Note that this won't build the pre-built binaries for ARM64 currently. Is there a way to do this on a x64 VM? For 32-bit we currently just run a 32-bit version of Node on the x64 VM to do the build.

dennisameling commented:
Just addressed your feedback, thanks for that 😊 

You can build for ARM64 on a x64 host by adding `--target_arch=arm64`, for example:

```
./node_modules/.bin/node-pre-gyp rebuild package testpackage --runtime=electron --target=11.2.3 --dist-url=https://electronjs.org/headers --target_arch=arm64
```

Just added that to the CircleCI config here: https://github.com/journeyapps/node-sqlcipher/pull/75/commits/ad13f7566151f96009caac6bccb67836277fda2c - **note that CircleCI doesn't run on this PR from my forked repo, so I can't currently test if this actually works.**

After having done that for all architectures, the output looks like this:

![image](https://user-images.githubusercontent.com/17739158/108739147-9bf1db00-7534-11eb-8e35-9fc867cb8847.png)

rkistner commented:
Thanks for all the work @dennisameling.

I've tested this, and seems to work well. Just had to make minor changes to the CircleCI config (see #76).

rkistner commented:
Released in v5.2.0.
